master/internal/rm/kubernetesrm/pods.go

```diff
@@ -897,10 +924,104 @@ func (p *pods) handleGetAgentsRequest(ctx *actor.Context) {
+	namespaceToQuota := make(map[string]k8sV1.ResourceQuota)
+
+	// Look up quotas for our resource pools' namespaces
+	for namespace := range p.namespaceToPoolName {
+		quotaList, err := p.quotaInterfaces[namespace].List(context.TODO(), metaV1.ListOptions{})
+		if k8serrors.IsNotFound(err) || quotaList == nil || len(quotaList.Items) != 1 {
+			// TODO: figure out how we want to handle multiple quotas per namespace?
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		namespaceToQuota[namespace] = quotaList.Items[0]
+	}
+
+	// If there's only one resource pool configured and it doesn't have a quota, summarize using the
+	// whole cluster.
+	if len(p.namespaceToPoolName) == 1 {
+		var namespaceOfPool string
+		for namespace := range p.namespaceToPoolName {
+			namespaceOfPool = namespace
+		}
+
+		// If there's no quota for our only resource pool's namespace
+		if _, ok := namespaceToQuota[namespaceOfPool]; !ok {
+			return p.summarizeCluster(ctx), nil
+		}
+	}
+
+	containers := p.containersPerResourcePool()
+	summaries := make(map[string]model.AgentSummary, len(p.namespaceToPoolName))
+	for namespace, poolName := range p.namespaceToPoolName {
+		quota, quotaExists := namespaceToQuota[namespace]
+		numContainers := containers[poolName]
+
+		var (
+			registeredTime time.Time
+			slots          map[string]model.SlotSummary
+		)
+		if quotaExists {
+			slots = make(map[string]model.SlotSummary)
+			registeredTime = quota.CreationTimestamp.Time
+
+			for resourceName, qty := range quota.Spec.Hard {
+				var deviceType device.Type
+				switch resourceName {
+				case k8sV1.ResourceCPU:
+					deviceType = device.CPU
+				case ResourceTypeNvidia:
+					deviceType = device.CUDA
+				default:
+					// We only care about CPU and GPU quotas for the slots summary
+					continue
+				}
+
+				// Each CPU and GPU in the quota will be counted as a slot here
+				one, decQty := inf.NewDec(1, 0), qty.AsDec()
+				for i := inf.NewDec(0, 0); i.Cmp(decQty) < 0; i.Add(i, one) {
+					id := fmt.Sprintf("%s/%s/%s", poolName, string(deviceType), i.String())
+
+					var container *cproto.Container
+					// Create a number of pseudo-containers in the summary equal to the number of
+					// running containers
+					if decNumContainers := inf.NewDec(int64(numContainers),
+						0); i.Cmp(decNumContainers) < 0 {
+						container = &cproto.Container{
+							ID:    cproto.ID(id),
+							State: "RUNNING",
+						}
+					}
+
+					slots[id] = model.SlotSummary{
+						ID:        id,
+						Device:    device.Device{Type: deviceType},
+						Enabled:   true,
+						Container: container,
+					}
+				}
+			}
+		}
+
+		summaries[poolName] = model.AgentSummary{
+			ID:             poolName,
+			RegisteredTime: registeredTime,
+			NumContainers:  numContainers,
+			ResourcePool:   poolName,
+			Slots:          slots,
+		}
+	}
+
+	return summaries, nil
+}
+
+func (p *pods) summarizeCluster(ctx *actor.Context) map[string]model.AgentSummary {
```

Comment:
Maybe this can now be called `summarizeClusterByNodes` or something to distinguish it from the quota-based logic.

Reply:
Ooh, that's a good name.